src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,124 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] arrayList;
+    private int size;
+
+    public ArrayList() {
+        arrayList = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == arrayList.length) {
+            resize(size);
+        }
+        arrayList[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index
+                    + " out of ArrayListSize");
+        }
+
+        if (size == arrayList.length) {
+            resize(arrayList.length * 2);
+        }
+
+        for (int i = size; i > index; i--) {
+            arrayList[i] = arrayList[i - 1];
+        }
 
+        arrayList[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        int arrayPlusListSize = size + list.size();
+        if (arrayPlusListSize > arrayList.length) {
+            resize(arrayPlusListSize);
+        }
 
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index
+                    + " out of ArrayListSize");
+        }
+        return (T) arrayList[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index
+                    + " out of ArrayListSize");
+        }
+        arrayList[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        int numMoved = size - index - 1;
+        if (index > size || index < 0 || numMoved < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index
+                    + " out of ArrayListSize");
+        }
+        final T removedElement = (T) arrayList[index];
+        System.arraycopy(arrayList, index + 1, arrayList, index, numMoved);
+        size--;
+        arrayList[size] = null;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = -1;
```

Comment:
rewrite this method without using such index approach. Return inside of a loop directly, throw exception without any checking after the loop. If we exited the loop, it means nothing is found. This is more common approach in such situations. Add exception message, include element in it